File: src/main/java/com/meetify/server/controller/PlaceRestController.java
package com.meetify.server.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.meetify.server.model.Id;
import com.meetify.server.model.Place;
import com.meetify.server.model.googleplace.GooglePlace;
import com.meetify.server.model.googleplace.Result;
import com.meetify.server.repository.PlaceRepository;
import com.meetify.server.repository.UserRepository;
import com.meetify.server.utils.HttpRequest;
import com.meetify.server.utils.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import javax.persistence.EntityManager;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

/**
 * Created by dmitry on 10/14/16.
 */

@RestController
@RequestMapping("/place")
public class PlaceRestController {
    private final UserRepository userRepository;
    private final PlaceRepository placeRepository;
    
    @Autowired
    public PlaceRestController(UserRepository userRepository, PlaceRepository placeRepository, @SuppressWarnings("SpringJavaAutowiringInspection") EntityManager entityManager) {
        this.userRepository = userRepository;
        this.placeRepository = placeRepository;
        //noinspection JpaQlInspection
        Object object = entityManager.createQuery("select max(place.id) from Place as place")
                .getResultList().get(0);
        Id id;
        if (object == null) {
            id = new Id((long) 0);
        } else {
            id = (Id) object;
        }
        Place.setCurrentId(id.getId() + 1);
    }
    
    @SuppressWarnings("ResultOfMethodCallIgnored")
    @RequestMapping("/getNearby")
    public ResponseEntity<?> getNearby(@RequestParam(value = "lat", defaultValue = "48.468488") String lat,
                                       @RequestParam(value = "lon", defaultValue = "35.049684") String lon,
                                       @RequestParam(value = "radius", defaultValue = "100") String radius) throws IOException {
        try {
            Double.parseDouble(lat);
            Double.parseDouble(lon);
            Double.parseDouble(radius);
        } catch (Exception e) {
            return new ResponseEntity<>(null, null, HttpStatus.FORBIDDEN);
        }
        ObjectMapper objectMapper = new ObjectMapper();
        GooglePlace googlePlace = objectMapper.readValue(StringUtils.makeString(HttpRequest.request(lat, lon, radius)), GooglePlace.class);
        googlePlace.setResults(sort(googlePlace.getResults()));
        googlePlace.getResults().forEach(result ->
                result.getPhotos().forEach(photo ->
                        photo.setPhotoReference(HttpRequest.photoRefToUrl(photo.getPhotoReference()))));
        return new ResponseEntity<>(googlePlace, null, HttpStatus.OK);
    }
    
    private List<Result> sort(List<Result> results) {
        List<Result> resultsChanged = new ArrayList<>();
        Optional.of(results).ifPresent(results1 -> results1.stream()
                .filter(result -> result.getPhotos() != null && result.getPhotos().size() > 0)
                .forEach(resultsChanged::add));
        return resultsChanged;
    }
    
    @RequestMapping(method = RequestMethod.GET)
    public ResponseEntity<?> get(@RequestParam(value = "id", defaultValue = "") String id) {
        return new ResponseEntity<>(placeRepository.findById(new Id(Long.parseLong(id))).orElse(null),
                null, HttpStatus.OK);
    }
    
    @RequestMapping(method = RequestMethod.POST)
    public ResponseEntity<?> post(@RequestParam(value = "name") String name,
                                  @RequestParam(value = "owner") String owner,
                                  @RequestParam(value = "allowed") String allowed) {
        Id id = new Id(Long.parseLong(owner));
        StringBuilder response = new StringBuilder();
        userRepository.findById(id).ifPresent(user1 -> {
            Place place = new Place(name, user1);
            if (!allowed.equals("")) {
                Arrays.stream(allowed.split(",")).forEach(s ->
                        userRepository.findById(new Id(Long.parseLong(s))).ifPresent(user -> {
                            user.getAllowed().add(place.getId());
                            place.getAllowed().add(user.getId());
                            userRepository.save(user);
                        })
                );
            }
            user1.getCreated().add(place);
            userRepository.save(user1);
            placeRepository.save(place);
            response.append(place.toString());
        });
        return new ResponseEntity<>(response.toString(), null, HttpStatus.OK);
    }
    
}

File: src/main/java/com/meetify/server/controller/PlaceRestController.kt
package com.meetify.server.controller

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.meetify.server.model.Id
import com.meetify.server.model.Location
import com.meetify.server.model.Place
import com.meetify.server.model.User
import com.meetify.server.model.googleplace.GooglePlace
import com.meetify.server.model.googleplace.Photo
import com.meetify.server.model.googleplace.Result
import com.meetify.server.repository.PlaceRepository
import com.meetify.server.repository.UserRepository
import com.meetify.server.utils.HttpRequest
import com.meetify.server.utils.JsonUtils
import com.meetify.server.utils.JsonUtils.mapper
import com.meetify.server.utils.MultiSearchable
import com.meetify.server.utils.StringUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.*
import java.util.*
import javax.persistence.EntityManager

/**
 * This class represents controller over places. It holds mapping '/place'.
 * Also, it is layer between client and Google Places Web API.
 * It's
 * @author  Dmitry Baynak
 * @version 0.0.1
 * @since   0.0.1
 * @param   userRepository  users repository.
 * @param   placeRepository places repository.
 * @param   entityManager   entity manager.
 * @constructor             Autowired by Spring.
 */
@RestController @RequestMapping("/place")
class PlaceRestController @Autowired constructor(private val userRepository: UserRepository,
                                                 private val placeRepository: PlaceRepository,
                                                 entityManager: EntityManager) : MultiSearchable<Place>() {

    /**
     * Returns Google Place, that downloaded from it's server.
     * It has pre-converted photo links and it doesn't contain any places without photos.
     * @param   locationJson    json representation of location near of which places are looking.
     * @return                  google place, which can be easily serialized with Jackson JSON library.
     */
    @ResponseBody @RequestMapping("/nearby", method = arrayOf(RequestMethod.GET))
    fun getNearby(@RequestParam(name = "location") locationJson: String): GooglePlace {
        val location = mapper.readValue(locationJson, Location::class.java)
        val googlePlace = jacksonObjectMapper().readValue(StringUtils.makeString(HttpRequest.request(location, "100")), GooglePlace::class.java)
        googlePlace.filterPhotos().forEach { result -> result.photos.convertRefs(); }
        return googlePlace
    }

    /**
     * Returns a list with information about places.
     * Unknown ids in [idsJson] list are ignored.
     * @param   idsJson json representation of collection of the ids of requested places.
     * @return          collection with places info.
     */
    @ResponseBody @RequestMapping(method = arrayOf(RequestMethod.GET))
    fun get(@RequestParam(name = "ids") idsJson: String): ArrayList<Place> {
        return get(placeRepository, JsonUtils.getList(idsJson))
    }

    /**
     * Method, that should be used create new users places.
     * Always create new place, even if place with exactly the same info is already present in database.
     * If owner of this place is not present in database, IllegalArgumentException is thrown.
     * If some ids in allowed are not associated with existing users, they are ignored.
     * @param   place   place, which should be created.
     * @return          place, that was created if case of success.
     */
    @ResponseBody @RequestMapping(method = arrayOf(RequestMethod.POST))
    fun post(@RequestBody place: Place): Place {
        userRepository.findById(place.id).orElseThrow { throw IllegalArgumentException("owner not found") }.apply {
            var usersToBeSaved: Set<User> = HashSet()
            place.allowed.forEach {
                userRepository.findById(it).ifPresent {
                    it.allowed += place.id
                    usersToBeSaved += it
                }
            }
            this.created += place.id
            usersToBeSaved += this
            userRepository.save(usersToBeSaved)
            placeRepository.save(place)
        }
        return place
    }

    /**
     * Dirty hack with places id. I wasn't able to create autoincrement embedded id, so it can be done with
     * simple searching max of existing ids, or, if it's not presented in database, set it as 1.
     */
    init {
        Place.currentId = 0
        Optional.ofNullable(entityManager
                .createQuery("select max(place.id) from Place as place")
                .resultList[0] as Id?).ifPresent {
            Place.currentId += it!!.id
        }
    }

    /**
     * Contains some useful methods, which are used to simplify code of some other methods.
     */
    companion object {

        /**
         * This function is used to convert simple photo references to downloadable links.
         */
        private fun List<Photo>.convertRefs() {
            forEach { photo -> photo.photoReference = HttpRequest.photoRefToUrl(photo.photoReference) }
        }

        /**
         * This function is used to discard places, that don't contain any photos.
         */
        private fun GooglePlace.filterPhotos(): List<Result> {
            results = results.filter { it.photos.size > 0 }
            return results
        }
    }
}
